### PR TITLE
Add support for rstream get/release record in the QUIC TLS layer

### PR DESCRIPTION
**Pull request by mattcaswell**
The QUIC TLS layer was taking an internal copy of rstream data while reading. The QUIC rstream code has recently been extended to enable a get/release model which avoids the need for this internal copy, so we use that instead.

We also make the "data" field of the TLS record layer "get_record" function "const". This improves consistency with the QUIC rstream implementation - and improves the abstraction between the TLS implementation and the abstract record layer. We should not expect that the TLS implementation should be able to change the underlying buffer. Future record layers may not expect that.

This is built on the commits in #19794, and is therefore draft until that PR has been merged.

### CODE REVIEW

**Comment by mattcaswell** [comment]
Now that #19794 has been merged, I've rebased this PR and taken it out of draft. It is now ready for review. Please take a look

**Comment by mattcaswell** [comment]
I added "test: exempted" for this. Is that the right label? Its a refactor, so no additional tests are required.

File: ssl/quic/quic_tls.c [hunk 1]
```diff
@@ -397,15 +389,30 @@ static int quic_read_record(OSSL_RECORD_LAYER *rl, void **rechandle,
...
-        return 0;
+        return OSSL_RECORD_RETURN_FATAL;
+    }
+
+    rl->recunreleased -= length;
+
+    if (rl->recunreleased > 0)
+        return OSSL_RECORD_RETURN_SUCCESS;
+
+    if (!rl->qtls->args.crypto_release_rcd_cb(rl->recread,
```
Comment (t8m) [inline ssl/quic/quic_tls.c:408]: This would not be necessary if you called the ossl_rstream_release_record only with the size of the data that was processed and subsequently called ossl_rstream_get_record again to get pointer to the remaining data. Not sure if this helps in the way how the record layer is called from the TLS handshake layer or not.

**Comment by mattcaswell** [comment]
Ping for review

File: ssl/record/methods/tls_common.c [hunk 2]
```diff
@@ -1132,15 +1132,32 @@ int tls_read_record(OSSL_RECORD_LAYER *rl, void **rechandle, int *rversion,
...
     }
 
+    if (rec->length < length) {
+        /* Should not happen */
+        RLAYERfatal(rl, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return OSSL_RECORD_RETURN_FATAL;
+    }
+
+    if ((rl->options & SSL_OP_CLEANSE_PLAINTEXT) != 0)
+        OPENSSL_cleanse(rec->data + rec->off, length);
```
Comment (t8m) [inline ssl/record/methods/tls_common.c:1153]: Hm... we do not cleanse the released data in rstream code. Should we? It would require adding some further arguments. That would be for separate PR though.

**Review by t8m — APPROVED** [review]

File: ssl/record/methods/tls_common.c (hunk 2 above)
Comment (mattcaswell) [reply ssl/record/methods/tls_common.c:1153]: We probably should do. But, yeah, a separate PR.

File: ssl/quic/quic_channel.c [hunk 3]
```diff
@@ -237,8 +238,10 @@ static int ch_init(QUIC_CHANNEL *ch)
     tls_args.s                          = ch->tls;
     tls_args.crypto_send_cb             = ch_on_crypto_send;
     tls_args.crypto_send_cb_arg         = ch;
-    tls_args.crypto_recv_cb             = ch_on_crypto_recv;
-    tls_args.crypto_recv_cb_arg         = ch;
+    tls_args.crypto_recv_rcd_cb         = ch_on_crypto_recv_record;
+    tls_args.crypto_recv_rcd_cb_arg     = ch;
+    tls_args.crypto_release_rcd_cb      = ch_on_crypto_release_record;
+    tls_args.crypto_release_rcd_cb_arg  = ch;
```
Comment (paulidale) [inline ssl/quic/quic_channel.c:244]: Six copies of the `ch` channel pointer seems excessive.

Comment (mattcaswell) [reply ssl/quic/quic_channel.c:244]: This is just continuing the existing pattern (we add one extra callback in this PR). Each callback has its own arg. I am ambivalent about this - I would prefer to say this outside of the scope of this PR.

**Comment by mattcaswell** [comment]
Still looking for a second approval on this.

**Comment by paulidale** [comment]
Merged!
